### PR TITLE
Build into local GOPATH (_output)

### PR DESCRIPTION
**Pull request by jawnsy**
Signed-off-by: Jonathan Yu <jawnsy@redhat.com>

### CODE REVIEW

**Comment by jawnsy** [comment]
@cyphar @mikebrow Can you give this a try and see if this is better than our current solution? This builds into a local `GOPATH` (`_output`, like kubernetes/origin, but customizable with `OUTDIR`).

This breaks the tool installation code, since it installs tools into _output rather than the system GOPATH. I want to follow up with a PR that will `go get` into `_output` unless the binaries are available on PATH, but will do that in a follow-up.

cc @runcom @mrunalp @nalind

**Comment by mrunalp** [comment]
@nalind The tests panic in copyimg. Could you take a look?

**Comment by mikebrow** [comment]
@jawnsy Needs rebase

**Comment by mikebrow** [comment]
After rebasing.. I got the following build error:
```
mike@mike-VirtualBox:~/go/src/github.com/kubernetes-incubator/cri-o$ sudo make
mkdir -p "_output/src/github.com/kubernetes-incubator/"
ln -s ../../../.. "_output/src/github.com/kubernetes-incubator/cri-o"
mkdir -p _output/bin
touch _output/.ok
go install \
	-tags "selinux seccomp  libdm_no_deferred_remove" \
	github.com/kubernetes-incubator/cri-o/cmd/ocid
go install github.com/kubernetes-incubator/cri-o/cmd/ocic
go install github.com/kubernetes-incubator/cri-o/cmd/kpod
make -C conmon
make[1]: Entering directory '/home/mike/go/src/github.com/kubernetes-incubator/cri-o/conmon'
cc -Wall -Wextra -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -c -o conmon.o conmon.c
cc -o conmon conmon.o -Wall -Wextra -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -lglib-2.0
make[1]: Leaving directory '/home/mike/go/src/github.com/kubernetes-incubator/cri-o/conmon'
make -C pause
make[1]: Entering directory '/home/mike/go/src/github.com/kubernetes-incubator/cri-o/pause'
cc -Os -Wall -Wextra -static   -c -o pause.o pause.c
cc -o pause pause.o -Os -Wall -Wextra -static 
strip pause
make[1]: Leaving directory '/home/mike/go/src/github.com/kubernetes-incubator/cri-o/pause'
make -C test/bin2img BUILDTAGS="selinux seccomp  libdm_no_deferred_remove"
make[1]: Entering directory '/home/mike/go/src/github.com/kubernetes-incubator/cri-o/test/bin2img'
go build -tags "selinux seccomp  libdm_no_deferred_remove" -o bin2img
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0x7e18e0, 0xc420010150)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
main.vendoredImportPath(0xc4201cd200, 0xc420321341, 0x1a, 0x3, 0x0)
	/usr/local/go/src/cmd/go/pkg.go:463 +0x66c
main.loadImport(0xc420321341, 0x1a, 0xc42014ff10, 0x62, 0xc4201cd200, 0xc4201e19e0, 0xc42014b680, 0x3, 0x4, 0x1, ...)
	/usr/local/go/src/cmd/go/pkg.go:333 +0x9ea
main.(*Package).load(0xc4201cd200, 0xc4201e19e0, 0xc420083880, 0x0, 0x0, 0x4)
	/usr/local/go/src/cmd/go/pkg.go:940 +0x12b6
main.loadImport(0xc42019e121, 0x23, 0xc420123d60, 0x44, 0xc420190000, 0xc4201e19e0, 0xc42019e540, 0x1, 0x1, 0x1, ...)
	/usr/local/go/src/cmd/go/pkg.go:374 +0x470
main.(*Package).load(0xc420190000, 0xc4201e19e0, 0xc420169880, 0x0, 0x0, 0xc)
	/usr/local/go/src/cmd/go/pkg.go:940 +0x12b6
main.loadImport(0x8488e5, 0x1, 0xc420014054, 0x44, 0x0, 0xc4201e19e0, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/cmd/go/pkg.go:374 +0x470
main.loadPackage(0x8488e5, 0x1, 0xc4201e19e0, 0x0)
	/usr/local/go/src/cmd/go/pkg.go:1658 +0x28f
main.packagesAndErrors(0xc420174aa0, 0x1, 0x0, 0xc420174a80, 0x20, 0x7f3a60)
	/usr/local/go/src/cmd/go/pkg.go:1703 +0x337
main.packagesForBuild(0xc42000c170, 0x0, 0x0, 0x893c68, 0xc420016040, 0x0)
	/usr/local/go/src/cmd/go/pkg.go:1719 +0x75
main.runBuild(0xa27440, 0xc42000c170, 0x0, 0x0)
	/usr/local/go/src/cmd/go/build.go:440 +0xd4
main.main()
	/usr/local/go/src/cmd/go/main.go:181 +0x624
Makefile:2: recipe for target 'bin2img' failed
make[1]: *** [bin2img] Error 2
make[1]: Leaving directory '/home/mike/go/src/github.com/kubernetes-incubator/cri-o/test/bin2img'
Makefile:58: recipe for target 'bin2img' failed
make: *** [bin2img] Error 2
``` 
Don't see this issue on master so presume some issue with the PR.

**Comment by jawnsy** [comment]
@mikebrow Sweet, thanks for testing. I don't see that problem locally, but it happens in Travis and I'm not sure how to fix it. But the good news is that things seem to build okay - got past the `go install` phase. :smile:

**Comment by mikebrow** [comment]
Obvious stuff to check .. setting GOBIN (override GOPATH/bin).. while also prepending the gopath with the absolute ./_output directory, were these needed? And how does that effect the go build that is failing. Maybe needs a -u option?

**Comment by runcom** [comment]
@nalind could you take a look at https://github.com/kubernetes-incubator/cri-o/pull/345#issuecomment-276524021? seems like a panic in bin2img or copyimg

**Comment by nalind** [comment]
The symlink in GOPATH is confusing the package resolution logic in the compiler.  @jawnsy can you cherry pick the most recent commit from https://github.com/nalind/cri-o/tree/subproject and see if that gets around it?

File: test/helpers.bash [hunk 1]
```diff
@@ -9,10 +9,14 @@ TESTDATA="${INTEGRATION_ROOT}/testdata"
 # Root directory of the repository.
 OCID_ROOT=${OCID_ROOT:-$(cd "$INTEGRATION_ROOT/../.."; pwd -P)}
 
+# Installation directory for Go binaries
+GOPATH=${GOPATH:-${OCID_ROOT}/cri-o/_output}
+GOBIN=${GOBIN:-${GOPATH}/bin}
```
Comment (nalind) [inline test/helpers.bash]: This breaks if `GOPATH` has more than one element in it.  Try `GOBIN=${GOBIN:-${GOPATH%%:*}/bin}` instead?

**Comment by jawnsy** [comment]
Superseded by #336

**Pull request closed without merging**